src/Catalog/PackageDownloader.cs

```diff
@@ -0,0 +1,96 @@
+    public class PackageDownloader
+    {
+        private const int BufferSize = 8192;
+
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
+
+        public PackageDownloader(HttpClient httpClient, ILogger logger)
+        {
+            _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task<Stream> DownloadAsync(Uri packageUri, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Attempting to download package from {PackageUri}...", packageUri);
+
+            Stream packageStream = null;
+            var stopwatch = Stopwatch.StartNew();
+
+            try
+            {
+                // Download the package from the network to a temporary file.
+                using (var response = await _httpClient.GetAsync(packageUri, HttpCompletionOption.ResponseHeadersRead))
+                {
+                    _logger.LogInformation(
+                        "Received response {StatusCode}: {ReasonPhrase} of type {ContentType} for request {PackageUri}",
+                        response.StatusCode,
+                        response.ReasonPhrase,
+                        response.Content.Headers.ContentType,
+                        packageUri);
+
+                    if (response.StatusCode == HttpStatusCode.NotFound)
+                    {
+                        return null;
+                    }
+                    if (response.StatusCode != HttpStatusCode.OK)
+                    {
+                        throw new InvalidOperationException($"Expected status code {HttpStatusCode.OK} for package download, actual: {response.StatusCode}");
+                    }
+
+                    using (var networkStream = await response.Content.ReadAsStreamAsync())
+                    {
+                        packageStream = new FileStream(
+                            Path.GetTempFileName(),
+                            FileMode.Create,
+                            FileAccess.ReadWrite,
+                            FileShare.None,
+                            BufferSize,
+                            FileOptions.DeleteOnClose | FileOptions.Asynchronous);
+
+                        await networkStream.CopyToAsync(packageStream, BufferSize, cancellationToken);
+                    }
+                }
+
+                packageStream.Position = 0;
+
+                stopwatch.Stop();
+
+                _logger.LogInformation(
+                    "Downloaded {PackageSizeInBytes} bytes in {DownloadElapsedTime} seconds for request {PackageUri}",
+                    packageStream.Length,
+                    stopwatch.Elapsed.TotalSeconds,
+                    packageUri);
+
+                return packageStream;
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(
+                    0,
```

Comment:
parameter name

Reply:
I'll do `(EventId)0`. This seems like a good middle ground to know what that `0` means and not having to name all further parameters.